### ISSUE TITLE
collector generator training presents same samples many times

### ISSUE THREAD

**Issue by alihaydaroglu** [AllenInstitute/deepinterpolation#69]
### Issue
![image](https://user-images.githubusercontent.com/15041006/141405098-b9adbab5-b6e8-4134-8161-de0603f94d1f.png)
Training runs as expected, but the loss patterns seem to suggest that the same frames are presented over and over again during training. This happens when using multiple tif files using the Collector Generator

### Steps to reproduce
Using the following training script: 
```python
import deepinterpolation as de
import sys
from shutil import copyfile
import os
from deepinterpolation.generic import JsonSaver, ClassLoader
import datetime
from typing import Any, Dict
import pathlib

trained_model_path = "C:\\ali\\data\\deepinterp\\models\\models_4\\"

n_pre_post_frames = 30
batch_size = 5
learning_rate=0.0001

if __name__=='__main__':
    # This is used for record-keeping
    now = datetime.datetime.now()
    run_uid = now.strftime("%Y_%m_%d_%H_%M")

    # Initialize meta-parameters objects
    training_param = {}
    network_param = {}
    generator_test_param = {}

    # An epoch is defined as the number of batches pulled from the dataset. Because our datasets are VERY large. Often, we cannot
    # go through the entirity of the data so we define an epoch slightly differently than is usual.
    steps_per_epoch = 5

    train_files = [
    'C:\\ali\\data\\deepinterp\\datasets\\AH001_2021-10-12_1\\plane1.tif',
    # 'C:\\ali\\data\\deepinterp\\datasets\\AH001_2021-10-12_1\\plane2.tif',
    # 'C:\\ali\\data\\deepinterp\\datasets\\AH001_2021-10-12_1\\plane3.tif',
    # 'C:\\ali\\data\\deepinterp\\datasets\\AH001_2021-10-12_1\\plane4.tif',
    'C:\\ali\\data\\deepinterp\\datasets\\AH001_2021-10-12_1\\plane5.tif',
    # 'C:\\ali\\data\\deepinterp\\datasets\\AH001_2021-10-12_1\\plane6.tif',
    # 'C:\\ali\\data\\deepinterp\\datasets\\AH001_2021-10-12_1\\plane7.tif',
    'C:\\ali\\data\\deepinterp\\datasets\\AH001_2021-10-12_1\\plane8.tif',
    'C:\\ali\\data\\deepinterp\\datasets\\AH002_2021-11-03_1\\plane1.tif',
    # 'C:\\ali\\data\\deepinterp\\datasets\\AH002_2021-11-03_1\\plane2.tif',
    # 'C:\\ali\\data\\deepinterp\\datasets\\AH002_2021-11-03_1\\plane3.tif',
    # 'C:\\ali\\data\\deepinterp\\datasets\\AH002_2021-11-03_1\\plane4.tif',
    'C:\\ali\\data\\deepinterp\\datasets\\AH002_2021-11-03_1\\plane5.tif',
    # 'C:\\ali\\data\\deepinterp\\datasets\\AH002_2021-11-03_1\\plane6.tif',
    # 'C:\\ali\\data\\deepinterp\\datasets\\AH002_2021-11-03_1\\plane7.tif',
    'C:\\ali\\data\\deepinterp\\datasets\\AH002_2021-11-03_1\\plane8.tif',
    ]
    test_file = 'C:\\ali\\data\\deepinterp\\datasets\\AH001_2021-10-29_1\\plane8.tif'

    generator_test_params = []
    for train_file in train_files:
        generator_param = {}
        # Those are parameters used for the main data generator
        generator_param["type"] = "generator"
        generator_param["steps_per_epoch"] = steps_per_epoch
        generator_param["name"] = "SingleTifGenerator"
        generator_param["pre_post_frame"] = n_pre_post_frames
        generator_param["train_path"] = train_file
        generator_param["batch_size"] = batch_size
        generator_param["start_frame"] = 0
        generator_param["end_frame"] = -1
        generator_param["pre_post_omission"] = 0

        generator_test_params.append(generator_param)

    # Those are parameters used for the Validation test generator. Here the test is done on the beginning of the data but
    # this can be a separate file
    generator_test_param["type"] = "generator"  # type of collection
    generator_test_param["name"] = "SingleTifGenerator"  # Name of object in the collection
    generator_test_param[
        "pre_post_frame"
    ] = n_pre_post_frames # Number of frame provided before and after the predicted frame
    generator_test_param["train_path"] = test_file
    generator_test_param["batch_size"] = 1
    generator_test_param["start_frame"] = 0
    generator_test_param["end_frame"] = 65
    generator_test_param[
        "pre_post_omission"
    ] = 0  # Number of frame omitted before and after the predicted frame
    generator_test_param[
        "steps_per_epoch"
    ] = -1  # No step necessary for testing as epochs are not relevant. -1 deactivate it.


    # Those are parameters used for the network topology
    network_param["type"] = "network"
    network_param["name"] = "unet_single_1024"  # Name of network topology in the collection

    # Those are parameters used for the training process
    training_param["type"] = "trainer"
    training_param["name"] = "transfer_trainer"
    training_param["run_uid"] = run_uid
    training_param["batch_size"] = generator_test_param["batch_size"]
    training_param["steps_per_epoch"] = steps_per_epoch
    training_param[
        "period_save"
    ] = 25  # network model is potentially saved during training between a regular nb epochs
    training_param["nb_gpus"] = 1
    training_param["apply_learning_decay"] = 0
    training_param[
        "nb_times_through_data"
    ] = 1  # if you want to cycle through the entire data. Two many iterations will cause noise overfitting
    training_param["learning_rate"] = learning_rate # 0.0001
    training_param["pre_post_frame"] = generator_test_param["pre_post_frame"]
    training_param["loss"] = "mean_absolute_error"
    training_param['use_multiprocessing'] = False
    training_param[
        "nb_workers"
    ] = 1  # this is to enable multiple threads for data generator loading. Useful when this is slower than training

    training_param[
            "model_path"
        ] = "C:\\ali\\data\\deepinterp\\models\\deep_interpolation_ai93_v1_1\\2019_09_11_23_32_unet_single_1024_mean_absolute_error_Ai93-0450.h5"

    training_param["model_string"] = (
        network_param["name"]
        + "_"
        + training_param["loss"]
        + "_"
        + training_param["run_uid"]
    )


    # Where do you store ongoing training progress
    jobdir = os.path.join(
        trained_model_path, training_param["model_string"] + "_" + run_uid,
    )
    training_param["output_dir"] = jobdir
    print(jobdir)
    try:
        os.mkdir(jobdir)
    except:
        print("folder already exists")

    path_training = os.path.join(jobdir, "training.json")
    json_obj = JsonSaver(training_param)
    json_obj.save_json(path_training)


    list_train_generator = []
    for local_index, indiv_generator in enumerate(generator_test_params):
        path_generator = os.path.join(
            jobdir, "generator" + str(local_index) + ".json")
        json_obj = JsonSaver(indiv_generator)
        json_obj.save_json(path_generator)
        generator_obj = ClassLoader(path_generator)
        train_generator = generator_obj.find_and_build()(path_generator)
        list_train_generator.append(train_generator)

    path_test_generator = os.path.join(jobdir, "test_generator.json")
    json_obj = JsonSaver(generator_test_param)
    json_obj.save_json(path_test_generator)

    path_network = os.path.join(jobdir, "network.json")
    json_obj = JsonSaver(network_param)
    json_obj.save_json(path_network)

    generator_test_obj = ClassLoader(path_test_generator)

    network_obj = ClassLoader(path_network)
    trainer_obj = ClassLoader(path_training)

    test_generator = generator_test_obj.find_and_build()(path_test_generator)

    network_callback = network_obj.find_and_build()(path_network)

    global_train_generator = de.generator_collection.CollectorGenerator(
        list_train_generator
    )

    training_class = trainer_obj.find_and_build()(
        global_train_generator, test_generator, path_training
    )

    training_class.run()

    training_class.finalize()

```

**Comment by alihaydaroglu** [comment]
From @jeromelecoq on Slack:

> the root of the issue is the following
the CollectorGenerator is missing this function :
def on_epoch_end(self):
        # We only increase index if steps_per_epoch
        # is set to positive value. -1 will force the generator
        # to not iterate at the end of each epoch
        if self.steps_per_epoch > 0:
            if self.steps_per_epoch * (self.epoch_index + 2) < self.__len__():
                self.epoch_index = self.epoch_index + 1
            else:
                # if we reach the end of the data, we roll over
                self.epoch_index = 0

**Comment by jeromelecoq** [comment]
This was fixed in the latest release. The collectorGenerator now also has an epoch_index.
